pixeltable/share/publish.py

```diff
@@ -165,45 +168,87 @@ def _download_bundle_from_s3(parsed_location: urllib.parse.ParseResult, bundle_f
+    )
+
+    session = requests.Session()
+    adapter = HTTPAdapter(max_retries=retry_strategy)
+    session.mount('https://', adapter)
+    return session
+
+
+def _upload_to_presigned_url(file_path: Path, url: str, max_retries: int = 3) -> requests.Response:
+    """Upload file with progress bar and retries"""
+    file_size = file_path.stat().st_size
+
     headers = {'Content-Length': str(file_size), 'Content-Type': 'application/octet-stream'}
-    for attempt in range(1, max_retries + 1):
-        try:
-            with open(bundle, 'rb') as f:
-                response = requests.put(presigned_url, data=f, headers=headers, stream=True, timeout=(60, 1800))
-                response.raise_for_status()
-                return
-        except (requests.exceptions.Timeout, requests.exceptions.SSLError, requests.exceptions.ConnectionError) as e:
-            if attempt == max_retries:
-                raise excs.Error(f'Upload failed after {max_retries} attempts: {e}') from e
-            time.sleep(1)  # wait before retrying
-            continue
-        except Exception as e:
-            raise excs.Error(f'Failed to upload : {e}') from e
-
-
-def _download_bundle_from_presigned_url(presigned_url: str, max_retries: int = 3) -> Path:
-    bundle_path = Path(Env.get().create_tmp_path())
-    for attempt in range(1, max_retries + 1):
-        try:
-            response = requests.get(presigned_url, stream=True, timeout=(60, 1800))
+
+    session = _create_retry_session(max_retries=max_retries)
+
+    try:
+        with (
+            open(file_path, 'rb') as f,
+            tqdm.wrapattr(f, 'read', total=file_size, desc='Uploading') as file_with_progress,
+        ):
+            response = session.put(
+                url,
+                data=file_with_progress,
+                headers=headers,
+                timeout=(60, 1800),  # 60 seconds to connect and 300 seconds for server response
+            )
             response.raise_for_status()
-            # Stream download to file
-            with open(bundle_path, 'wb') as f:
-                for chunk in response.iter_content(chunk_size=8192):
+            return response
+    finally:
+        session.close()
+
+
+def _download_from_presigned_url(
+    url: str, output_path: Path, headers: Optional[Dict[str, str]] = None, max_retries: int = 3
+) -> None:
+    """Download file with progress bar and retries"""
+    session = _create_retry_session(max_retries=max_retries)
+
+    try:
+        # Stream download with progress
+        response = session.get(
+            url, headers=headers, stream=True, timeout=(60, 300)
+        )  # 60 seconds to connect and 300 seconds for server response
+        response.raise_for_status()
+
+        total_size = int(response.headers.get('content-length', 0))
+
+        with (
+            open(output_path, 'wb') as f,
+            tqdm(total=total_size, unit='B', unit_scale=True, desc='Downloading') as pbar,
```

Comment:
Use the same tqdm configuration as for S3 transfers to ensure a consistent experience.